Sources/TensorFlow/Operators/LinearAlgebra.swift

```diff
@@ -266,3 +266,131 @@ public extension Tensor where Scalar: TensorFlowFloatingPoint {
         return (s, u, v)
     }
 }
+
+
+// MARK: Solvers
+
+
+/// Solves optionally batched systems of linear equations with upper or lower triangular
+/// matrices by backsubstitution.
+/// Returns solution to the system `A x = b`. Shape of return matches `b`.
+///
+/// - Parameters:
+///     - matrix: A batched matrix tensor.
+///     - rhs: A batched vector tensor.
+///     - lower: Boolean option indicating whether the innermost matrices in
+///         matrix are lower or upper triangular. Defaults to true.
+///     - adjoint: Boolean option indicating whether to solve with matrix or
+///         its (block-wise) adjoint. Defaults to False.
+/// - Precondition: `matrix` must be a tensor with shape `[..., M, M]`.
+/// - Precondition: `rhs` must be a tensor with shape `[..., M, K]`.
+@inlinable
+@differentiable
+public func triangularSolve<T: TensorFlowFloatingPoint>(
+    matrix: Tensor<T>,
+    rhs: Tensor<T>,
+    lower: Bool = true,
+    adjoint: Bool = false
+) -> Tensor<T> {
+    precondition(matrix.rank >= 2, "The matrix tensor must have at least rank 2.")
+    precondition(rhs.rank >= 2, "The rhs tensor must have at least rank 2.")
+    if matrix.rank < rhs.rank {
```

Comment:
I wonder if you used some reference implementation and tests for `triangularSolve` and `_vjpTriangularSolve` in this PR? I'm particularly curious about the `extractLeadingDims` helper function.

Does `triangularSolve` match [`tf.linalg.triangular_solve`](https://www.tensorflow.org/api_docs/python/tf/linalg/triangular_solve) in behavior, or does it have different/additional functionality?

Reply:
The `trisolve` from TensorFlow doesn't work for batched cases. https://github.com/tensorflow/tensorflow/issues/26204#issuecomment-498989256. I will make same contribution to the TF2.0 soon as well.

Reply:
@dan-zheng, the trisolve gradient you can find here https://github.com/tensorflow/tensorflow/blob/master/tensorflow/python/ops/linalg_grad.py#L607